File: 05/gemini-3.0-pro/rust/src/main.rs
use std::fs;

struct Range {
    start: i64,
    end: i64,
}

fn main() {
    let input = fs::read_to_string("input.txt")
        .expect("Failed to read input")
        .trim()
        .replace("\r\n", "\n");

    let parts: Vec<&str> = input.split("\n\n").collect();
    let ranges_lines = parts[0].lines();
    let ids_lines = parts[1].lines();

    let mut ranges = Vec::new();
    for line in ranges_lines {
        let split: Vec<&str> = line.split('-').collect();
        let start: i64 = split[0].parse().unwrap();
        let end: i64 = split[1].parse().unwrap();
        ranges.push(Range { start, end });
    }

    let mut part1: i64 = 0;
    for line in ids_lines {
        let id: i64 = line.parse().unwrap();
        let mut is_fresh = false;
        for range in &ranges {
            if id >= range.start && id <= range.end {
                is_fresh = true;
                break;
            }
        }
        if is_fresh {
            part1 += 1;
        }
    }

    println!("Day 05 Part 1: {}", part1);

    // Part 2
    ranges.sort_by(|a, b| a.start.cmp(&b.start));

    let mut merged_ranges: Vec<Range> = Vec::new();

    if !ranges.is_empty() {
        let mut curr_start = ranges[0].start;
        let mut curr_end = ranges[0].end;

        for i in 1..ranges.len() {
            let next_start = ranges[i].start;
            let next_end = ranges[i].end;

            if next_start <= curr_end + 1 {
                if next_end > curr_end {
                    curr_end = next_end;
                }
            } else {
                merged_ranges.push(Range { start: curr_start, end: curr_end });
                curr_start = next_start;
                curr_end = next_end;
            }
        }
        merged_ranges.push(Range { start: curr_start, end: curr_end });
    }

    let mut part2: i64 = 0;
    for range in merged_ranges {
        part2 += range.end - range.start + 1;
    }

    println!("Day 05 Part 2: {}", part2);
}

File: 05/gemini-3.0-pro/go/main.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Range struct {
	Start int64
	End   int64
}

func main() {
	data, _ := os.ReadFile("input.txt")
	input := strings.TrimSpace(string(data))
	input = strings.ReplaceAll(input, "\r\n", "\n")
	parts := strings.Split(input, "\n\n")

	rangesLines := strings.Split(parts[0], "\n")
	idsLines := strings.Split(parts[1], "\n")

	var ranges []Range
	for _, line := range rangesLines {
		if line == "" {
			continue
		}
		split := strings.Split(line, "-")
		start, _ := strconv.ParseInt(split[0], 10, 64)
		end, _ := strconv.ParseInt(split[1], 10, 64)
		ranges = append(ranges, Range{Start: start, End: end})
	}

	var part1 int64 = 0
	for _, line := range idsLines {
		if line == "" {
			continue
		}
		id, _ := strconv.ParseInt(line, 10, 64)
		isFresh := false
		for _, r := range ranges {
			if id >= r.Start && id <= r.End {
				isFresh = true
				break
			}
		}
		if isFresh {
			part1++
		}
	}

	fmt.Printf("Day 05 Part 1: %d\n", part1)

	// Part 2
	sort.Slice(ranges, func(i, j int) bool {
		return ranges[i].Start < ranges[j].Start
	})

	var mergedRanges []Range
	if len(ranges) > 0 {
		currStart := ranges[0].Start
		currEnd := ranges[0].End

		for i := 1; i < len(ranges); i++ {
			nextStart := ranges[i].Start
			nextEnd := ranges[i].End

			if nextStart <= currEnd+1 {
				if nextEnd > currEnd {
					currEnd = nextEnd
				}
			} else {
				mergedRanges = append(mergedRanges, Range{Start: currStart, End: currEnd})
				currStart = nextStart
				currEnd = nextEnd
			}
		}
		mergedRanges = append(mergedRanges, Range{Start: currStart, End: currEnd})
	}

	var part2 int64 = 0
	for _, r := range mergedRanges {
		part2 += (r.End - r.Start + 1)
	}

	fmt.Printf("Day 05 Part 2: %d\n", part2)
}
